File: examples/polyglot/typescript/test-example-board.ts
/**
 * End-to-end test of the generated TypeScript typed client.
 *
 * Run with:
 *   cd python && uv run jmp shell --exporter-config ../examples/polyglot/exporter.yaml \
 *     -- npx tsx ../examples/polyglot/typescript/test-example-board.ts
 */

import * as net from "net";
import { createExampleBoardDevice } from "./gen/src/testing";

async function main() {
  const ctx = await createExampleBoardDevice();
  const device = ctx.device;

  console.log("Device created successfully\n");

  // -- Power tests --
  console.log("--- Power Tests ---");

  await device.power.on();
  console.log("power.on(): OK");

  await device.power.off();
  console.log("power.off(): OK");

  for await (const reading of device.power.read()) {
    console.log(`power.read(): voltage=${reading.voltage}, current=${reading.current}`);
    break; // Just verify the first reading
  }

  // -- Storage Mux tests --
  console.log("\n--- Storage Mux Tests ---");

  await device.storage.host();
  console.log("storage.host(): OK");

  await device.storage.dut();
  console.log("storage.dut(): OK");

  await device.storage.off();
  console.log("storage.off(): OK");

  // -- Network echo test --
  console.log("\n--- Network Tests ---");
  if (device.network) {
    const { address, port } = await device.network.connectTcp();
    console.log(`network.connectTcp(): listening on ${address}:${port}`);

    const socket = net.createConnection(port, address);
    await new Promise<void>((resolve) => socket.once("connect", resolve));

    socket.write("hello");
    const data = await new Promise<Buffer>((resolve) => socket.once("data", resolve));
    const received = data.toString();
    console.log(`network echo: sent "hello", received "${received}"`);
    if (received !== "hello") {
      throw new Error(`Expected "hello" but got "${received}"`);
    }
    console.log("network echo: OK");

    socket.destroy();
    device.network.close();
  } else {
    console.log("network driver is undefined (optional — skipping echo test)");
  }

  console.log("\n=== All TypeScript tests PASSED! ===");
  ctx.close();
  process.exit(0);
}

main().catch((err) => {
  console.error("Test failed:", err);
  process.exit(1);
});

File: examples/polyglot/rust/tests/examples/test_example_board.rs
//! End-to-end smoke test of the Rust typed client against a real exporter.
//!
//! Run with:
//!   cd python && uv run jmp shell --exporter-config ../examples/polyglot/exporter.yaml \
//!     -- cargo run -p jumpstarter-example-board-tests --example test_example_board

use jumpstarter_client::ExporterSession;
use jumpstarter_example_board::devices::example_board::ExampleBoardDevice;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let session = ExporterSession::from_env().await?;
    println!("Connected to exporter");

    let drivers: Vec<_> = session.report().instances().iter()
        .filter_map(|d| d.labels().get("jumpstarter.dev/name").cloned())
        .collect();
    println!("Discovered drivers: {:?}", drivers);

    let mut device = ExampleBoardDevice::new(&session);

    println!("\n--- Power Tests ---");
    device.power.on().await?;
    println!("power.on(): OK");
    device.power.off().await?;
    println!("power.off(): OK");

    let mut stream = device.power.read().await?;
    if let Some(reading) = stream.message().await? {
        println!("power.read(): voltage={}, current={}", reading.voltage, reading.current);
    }

    println!("\n--- Storage Mux Tests ---");
    device.storage.host().await?;
    println!("storage.host(): OK");
    device.storage.dut().await?;
    println!("storage.dut(): OK");
    device.storage.off().await?;
    println!("storage.off(): OK");

    println!("\n--- Optional Network ---");
    match &device.network {
        Some(_) => println!("network driver available"),
        None => println!("network driver is None (optional)"),
    }

    println!("\n=== All Rust tests PASSED! ===");
    Ok(())
}
